tips/TIP-0018/tip-0018.md

```diff
@@ -0,0 +1,3755 @@
+address of `Alias A` introduces a circular dependency and is not well-defined. By requiring the *Unlock Blocks* to be
+ordered as described above, a transaction consuming `Alias A` as well as `Alias B` can never be valid as there would
+always need to be one *Alias Unlock Block* referencing a greater index.
+
+#### Alias Unlock Block Syntactic Validation
+
+ - It must hold that 0 ≤ `Alias Reference Unlock Index` ≤ `Max Inputs Count`.
+
+#### Alias Unlock Block Semantic Validation
+
+ - The address of the unlocking condition of the input being unlocked must be an <i>Alias Address</i>.
+ - The index `i` of the <i>Alias Unlock Block</i> is the index of the input in the transaction that it unlocks.
+   `Alias Reference Unlock Index` must be < `i`.
+ - `Alias Reference Unlock Index` defines a previous input of the transaction and its unlock block. This input must
+   be an <i>Alias Output</i> with `Alias ID` that refers to the <i>Alias Address</i> being unlocked.
+ - The referenced <i>Alias Output</i> must be unlocked for state transition.
+
+### NFT Locking & Unlocking
+
+`NFT ID` field is functionally equivalent to `Alias ID` of an alias output. It is generated the same way, but it can
+only exist in NFT outputs. Following the same analogy as for alias addresses, NFT addresses are iota addresses that are
+controlled by whoever owns the NFT output itself.
+
+Outputs that are locked under `NFT Address` can be unlocked by unlocking the NFT output in the same transaction that
+defines `NFT Address`, that is, the NFT output where `NFT ID = NFT Address`.
+
+An **NFT Unlock Block** looks and behaves like an <i>Alias Unlock Block</i>, but the referenced input at the index must
+be an NFT output with the matching `NFT ID`.
+
+<details>
+<summary>NFT Unlock Block</summary>
+<blockquote>
+        Points to the unlock block of a consumed NFT output.
+</blockquote>
+</details>
+
+<table>
+    <tr>
+        <td><b>Name</b></td>
+        <td><b>Type</b></td>
+        <td><b>Description</b></td>
+    </tr>
+    <tr>
+        <td>Unlock Block Type</td>
+        <td>uint8</td>
+        <td>
+            Set to <strong>value 3</strong> to denote a <i>NFT Unlock Block</i>.
+        </td>
+    </tr>
+    <tr>
+        <td>NFT Reference Unlock Index</td>
+        <td>uint16</td>
+        <td>
+            Index of input and unlock block corresponding to an NFT output.
+        </td>
+    </tr>
+</table>
+
+An *Alias Unlock Block* is only valid if the input in the transaction at index `NFT Reference Unlock Index` is the NFT
+output with the same `NFT ID` as the `Address` field of the to-be unlocked output.
+
+If the i-th *Unlock Block* of a transaction is an *NFT Unlock Block* and has `NFT Reference Unlock Index` set to k, it
+must hold that i > k. Hence, an <i>NFT Unlock Block</i> can only reference an *Unlock Block* at a smaller index.
+
+#### NFT Unlock Block Syntactic Validation
+
+- It must hold that 0 ≤ `NFT Reference Unlock Index` ≤ `Max Inputs Count`.
```

Comment:
Same, shouldn't this be `< Max Inputs Count` ?

Reply:
yes